active/0000-prebuild-addons.md

```diff
@@ -0,0 +1,95 @@
+# Summary
+
+Allow Ember Addons to be "pre-built" such that they produce assets which can be included more efficiently in the build process of a consuming application.
+
+# Motivation
+
+When working on Ember-CLI applications, it is common to include many addons; this is one of the great benefits of the Ember ecosystem. However, in our current system, addons are built with the application they are included in. This means that for each addon you include in your application, you incur a non-trivial impact to your application's build performance.
+
+Since most addon consumers never customize an addon's build or change its source code, it would be beneficial to provide pre-built versions of their assets to improve build performance. Doing so will benefit all applications and allow larger applications, especially those being developed with Engines (which are packaged as addons), to scale to truly ambitious levels.
+
+# Detailed design
+
+## Pre-builds for Addons
+
+A new command will be introduced for use inside of addons only and used as part of the publishing process:
+
+```bash
+ember pre-build
+```
+
+This command will generate assets into a directory named `prebuilt` in the root of the addon. Within `prebuilt` will be a series of directories representing prebuilt assets for different combinations of targets and environments. Since targets can be non-primitive values, they will be hashed for the corresponding directory path.
+
+```bash
+cool-addon/
+  |-prebuilt/
+    |-48e380a165c417253512a904d6b5cf2b-prod/
+    |-48e380a165c417253512a904d6b5cf2b-dev/
+```
+
+Within each of the prebuilt asset directories, there will be directories which contain the results of each `treeFor*` hook in the addon prior to being merged with a consuming application.
+
+```bash
+cool-addon/prebuilt/48e380a165c417253512a904d6b5cf2b-prod/
+  |-addon/
+  |-addon-templates/
+  |-addon-test-support/
+  |-app/
+  |-public/
+  |-styles/
+  |-templates/
+  |-test-support/
+  |-vendor/
+```
+
+If any of the above folders would be empty, then they will be excluded from the output.
+
+### Specifying Targets and Environments
+
+Pre-built asset bundles will be generated for Ember-CLI's default target as well as development and production environments. Since custom targets and environments are likely to not have high reuse, they will be excluded from pre-builds to avoid penalizing all consumers during package installation.
+
+## Consuming Pre-builds
+
+During the build of consuming applications, if the target and environment information match one of the prebuilt asset bundles for a given addon, then those assets will be used by the build pipeline in place of the various `treeFor*` hooks that would normally execute for that addon.
+
+If a consumer wishes to build the addons from scratch with their application, they will be able to specify an `ignorePrebuiltAddons` flag in their `ember-cli-build.js`, like so:
+
+```js
+const app = new EmberApp(defaults, {
+  ignorePrebuiltAddons: true
```

Comment:
Good suggestion and this could feed nicely into the opt-back-in story for app's with non-default build targets.